addons/delivery_shipper/models/delivery_carrier.py

```diff
@@ -0,0 +1,245 @@
+        comodel_name='res.partner',
+        domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]",
+        default=lambda self: self.env.company.partner_id,
+    )
+    shipper_3pl_partner = fields.Selection([
+        ('57', 'J&T - Regular'),
+        ('228', 'Ninja Xpress - Regular'),
+        ('349', 'SAP - Regular'),
+        ('58', 'SiCepat - Regular'),
+        ('562', 'Anteraja - Regular'),
+        ('1', 'JNE - Regular'),
+        ('564', 'Anteraja - Express'),
+        ('59', 'SiCepat - Express'),
+        ('2', 'JNE - Express'),
+    ], string="Shipper 3PL Partners")
+    shipper_3pl_partner_id = fields.Integer(compute='_compute_shipper_3pl_partner_id', store=True)
+
+    @api.depends('shipper_3pl_partner')
+    def _compute_shipper_3pl_partner_id(self):
+        for record in self:
+            record.shipper_3pl_partner_id = int(record.shipper_3pl_partner) if record.shipper_3pl_partner else 0
+
+    # Shipping Carrier Methods
+    # TODO: Add an option for using geolocation instead; maybe
+    def shipper_rate_shipment(self, order):
+        shipper = self._get_shipper()
+        origin_area_id = self._shipper_get_area_id(origin=self.shipper_origin_address)
+        destination_area_id = self._shipper_get_area_id(order=order)
+
+        if not origin_area_id:
+            raise UserError(_('Cannot find origin address on Shipper. Please recheck the contact\'s address or postcode.'))
+        elif not destination_area_id:
+            raise UserError(_('Cannot find destination address on Shipper. Please recheck the contact\'s address or postcode.'))
+
+        rates = shipper._rate_shipment(
+            self._shipper_get_package_information(order=order),
+            origin_area_id,
+            destination_area_id,
+            order=order,
+        )
+
+        # Get the selected partner details
+        selected_partner = self.shipper_3pl_partner
+        selected_id = self.shipper_3pl_partner_id
+        selection_dict = dict(self.fields_get(['shipper_3pl_partner'])['shipper_3pl_partner']['selection'])
+
+        if selected_partner:
+            selected_label = selection_dict[selected_partner]
+            selected_name, selected_rate_type = selected_label.split(" - ", 1)
+        else:
+            selected_name, selected_rate_type = None, None
+
+        filtered_rates = [
+            rate for rate in rates
+            if rate['rate_id'] == selected_id or (
+                selected_name.lower() == rate['carrier_name'].lower() and
+                selected_rate_type.lower() == rate['rate_type'].lower()
+            )
+        ]
+        if not filtered_rates:
+            raise UserError(_('No rates can be found. Please check the package dimension or Shipper configuration'))
+
+        if filtered_rates[0]:
+            if filtered_rates[0]['rate_id'] != selected_id:
+                self.shipper_3pl_partner_id = filtered_rates[0]['rate_id']
```

Comment:
I don't understand this part. The user selects a "3PL" partner, but the partner used will be chosen from the first rate returned. Can't you provide the rate for the selected partner?